custom_components/tesla_custom/sensor.py

```diff
@@ -1,112 +1,455 @@
+    @property
+    def native_unit_of_measurement(self) -> str:
+        """Return distance units from car settings."""
+        return self._car.gui_distance_units
+
     @property
-    def state_class(self) -> str | None:
-        """Return the state_class of the device."""
-        try:
-            return (
-                self.tesla_device.state_class
-                if self.tesla_device.state_class in STATE_CLASSES
-                else None
+    def native_value(self) -> int:
+        """Return charge rate."""
+        charge_rate = self._car.charge_rate
+
+        # If we don't have anything, just return None.
+        if charge_rate is None:
+            return charge_rate
+
+        if self.unit_of_measurement == "km/hr":
+            charge_rate = round(
+                convert(charge_rate, LENGTH_MILES, LENGTH_KILOMETERS), 2
             )
-        except AttributeError:
-            return None
+
+        return charge_rate
 
     @property
     def extra_state_attributes(self):
-        """Return the state attributes of the device."""
-        attr = self._attributes.copy()
-        if self.tesla_device.type == "charging rate sensor":
-            attr.update(
-                {
-                    "time_left": self.tesla_device.time_left,
-                    "added_range": self.tesla_device.added_range,
-                    "charge_energy_added": self.tesla_device.charge_energy_added,
-                    "charge_current_request": self.tesla_device.charge_current_request,
-                    "charge_current_request_max": self.tesla_device.charge_current_request_max,
-                    "charger_actual_current": self.tesla_device.charger_actual_current,
-                    "charger_voltage": self.tesla_device.charger_voltage,
-                    "charger_power": self.tesla_device.charger_power,
-                }
+        """Return device state attributes."""
+        added_range = self._car.charge_miles_added_ideal
+
+        if self._car.gui_range_display == "Rated":
+            added_range = self._car.charge_miles_added_rated
+
+        if self.unit_of_measurement == "km/hr":
+            added_range = round(
+                convert(added_range, LENGTH_MILES, LENGTH_KILOMETERS), 2
             )
-        return attr
+
+        data = {
+            "time_left": self._car.time_to_full_charge,
+            "added_range": added_range,
+            "charge_energy_added": self._car.charge_energy_added,
+            "charge_current_request": self._car.charge_current_request,
+            "charge_current_request_max": self._car.charge_current_request_max,
+            "charger_actual_current": self._car.charger_actual_current,
+            "charger_voltage": self._car.charger_voltage,
+            "charger_power": self._car.charger_power,
+            "charger_phases": self._car.charger_phases,
+            "charge_limit_soc": self._car.charge_limit_soc,
+        }
+        self.attrs.update(data)
+        return self.attrs
```

Comment:
Any idea why we're using the `update` method here but not for other `extra_state_attributes`? All of this data should always be cached in teslajsonpy, so just returning the dictionary should work.

Reply:
HA used to require that attributes were copied from the underlying library so if the underlying library changed attributes, it wouldn't break HA. It was sorta stupid if you maintained both since you could control both but you had to maintain double the work. For the rewrite, you can probably make it sensible.

Reply:
It looks like this is something @Megabytemb intentionally added with his rewrite because in the current version it simply returns `attrs`. I just can't think of a reason why we would need to use `update` but I'm sure he had a reason. Just wish I knew what it was. 😄

Reply:
`update` just changes the keys that are defined in the new data and leaves other keys intact.  It's a shortcut to updating each key individually.

Reply:
Yeah that part makes sense, but the entire dictionary (`data`) is being defined/set each time and every value is cached by teslajsonpy. Usually when I've used `update` is when I'm trying to just update certain keys and want to retain the keys that already exist but are not being updated.

I realize this is super minor but just trying to go through and do some clean up/make sure everything makes sense. This is just one area that seems inconsistent across the different entities. Appreciate the help!